packages/markdown-pdf/src/PdfTransformer.js

```diff
@@ -83,87 +87,81 @@ class PdfTransformer {
      * @param {string} [format] result format, defaults to 'concerto'. Pass
      * 'json' to return the JSON data.
      * @param {object} [options] - the PDF parsing options
-     * @param {number} [options.paragraphVerticalOffset] - the vertical offset used to detect pararaphs (defaults to 1)
-     * @param {boolean} [options.preservePages] - whether to preserve page breaks (defaults to true)
+     * @param {number} [options.paragraphVerticalOffset] - the vertical offset used to detect
+     * pararaphs as a multiple of the line height (defaults to 2)
+     * @param {boolean} [options.preservePages] - whether to preserve PDF page breaks (defaults to true)
      * @returns {promise} a Promise to the CiceroMark DOM
      */
-    async toCiceroMark(input, format = 'concerto', options = { paragraphVerticalOffset: 1, preservePages: true }) {
-        return new Promise( (resolve, reject) => {
-            const pdfParser = new PDFParser(null, false);
-            const errorCallback = (errData) => reject(`PDF parsing failed with error ${errData.parserError}`);
-            const conversionCallback = (pdfData) => {
-
-                const document = {
-                    $class : 'org.accordproject.commonmark.Document',
-                    xmlns : pdfData.formImage.Id.Name,
-                    nodes : []
+    async toCiceroMark(input, format = 'concerto', options = { paragraphVerticalOffset: 2, preservePages: true }) {
+
+        let loadingTask = pdfjsLib.getDocument(input.buffer);
+
+        const doc = await loadingTask.promise;
+        let numPages = doc.numPages;
+        const metadata = await doc.getMetadata();
+
+        const pages = [];
+        for( let n=1; n <= numPages; n++) {
+            const page = await doc.getPage(n);
+            const content = await page.getTextContent({
+                normalizeWhitespace: true,
+                disableCombineTextItems: true,
+            });
+
+            let currentPara = null;
+            let lastY = 0;
+            const result = {
+                nodes: []
+            };
+
+            content.items.forEach( text => {
+                const tx = text.transform;
+                const textY = tx[5];
+                const height = text.height;
+                const newPara = Math.abs(lastY - textY) > (height * options.paragraphVerticalOffset);
+
+                if(!currentPara || newPara) {
+                    currentPara = {
+                        $class : 'org.accordproject.commonmark.Paragraph',
+                        nodes : []
+                    };
+                    result.nodes.push(currentPara);
+                }
+
+                const textNode = {
+                    $class : 'org.accordproject.commonmark.Text',
+                    text : text.str.replace(/(?:\r\n|\r|\n)/g, ' ')
```

Comment:
Would preserving new lines as soft-breaks be useful or harmful in this context?

Reply:
I think they add a lot of noise (and size) to the CommonMark JSON, because all line-endings in the PDF would become Softbreaks.